src/measure_theory/probability_mass_function.lean

```diff
@@ -148,6 +151,42 @@ def of_multiset (s : multiset α) (hs : s ≠ 0) : pmf α :=
 def of_fintype [fintype α] (f : α → ℝ≥0) (h : ∑ x, f x = 1) : pmf α :=
 ⟨f, h ▸ has_sum_sum_of_ne_finset_zero (by simp)⟩
 
+/-- Given a `f` with non-zero sum, we get a `pmf` by normalizing `f` by its `tsum` -/
+def normalize (f : α → ℝ≥0) (hf0 : tsum f ≠ 0) : pmf α :=
+⟨λ a, f a * (∑' x, f x)⁻¹,
+  (mul_inv_cancel hf0) ▸ has_sum.mul_right (∑' x, f x)⁻¹
+    (not_not.mp (mt tsum_eq_zero_of_not_summable hf0 : ¬¬summable f)).has_sum⟩
+
+lemma normalize_apply (f : α → ℝ≥0) {hf0 : tsum f ≠ 0} {a : α} :
+  (normalize f hf0) a = f a * (∑' x, f x)⁻¹ := rfl
+
+/-- Create new `pmf` by filtering on a set with non-zero measure and normalizing -/
+def filter (p : pmf α) (s : set α) (h : ∃ a ∈ s, p a ≠ 0) : pmf α :=
+pmf.normalize (s.indicator p) $ nnreal.tsum_indicator_ne_zero p.2.summable h
+
+lemma filter_apply (p : pmf α) {s : set α} {h : ∃ a ∈ s, p a ≠ 0} {a : α} :
```

Comment:
The `h` and `a` parameters should be explicit here, as they can not be inferred from anything. If you want to rewrite, this is not a problem (`rw filter_apply` will work just fine), but if you want to do something else being able to specify the parameters may help. Same thing in all the other lemmas: a parameter should be implicit only when it can be inferred from subsequent parameters.

Reply:
I had assumed it was also fine if the parameter could be inferred from the return type, but I see how that could cause issues if multiple patterns in the target match the return type. I've changed these all to be explicit now.

Reply:
I had assumed it was also fine if the parameter could be inferred from the return type, but I see how that could cause issues if multiple patterns in the target match the return type. I've changed these all to be explicit now.